File: src/arraysandstrings/SquaresOfASortedArray.java
package arraysandstrings;

/**
 * Given an integer array nums sorted in non-decreasing order, return an array of the squares of each number sorted in non-decreasing order.
 * Example:
 * Input: nums = [-7,-3,2,3,11]
 * Output: [4,9,9,49,121]
 * Follow up: Squaring each element and sorting the new array is very trivial, could you find an O(n) solution using a different approach?
 */
public class SquaresOfASortedArray {
    public static int[] sortedSquares(int[] a) {
        int n = a.length;
        int[] r = new int[n];
        int left = 0, right = n - 1, pos = n - 1;

        while (left <= right) {
            int leftSquare = a[left] * a[left];
            int rightSquare = a[right] * a[right];

            if(leftSquare > rightSquare) {
                r[pos] = leftSquare;
                left++;
            } else {
                r[pos] = rightSquare;
                right--;
            }
            pos--;
        }

        return r;
    }
}

File: src/main/kotlin/arraysandstrings/SquaresOfASortedArray.kt
package arraysandstrings

class SquaresOfASortedArray {
    /**
     * Given an integer array a sorted in non-decreasing order, return an array of the squares of each number
     * sorted in non-decreasing order.
     * Approach: Two pointers
     *
     * @param a input array, sorted in non-decreasing order, example: [-7,-3,2,3,11]
     * @return array of squares, sorted in non-decreasing order, example: [4,9,9,49,121]
     */
    fun sortedSquares(a: IntArray): IntArray {
        val n = a.count()
        val r = IntArray(n)
        var i = 0
        var j = n - 1
        var k = n - 1

        while (i <= j) {
            val ls = a[i] * a[i]
            val rs = a[j] * a[j]

            if (ls > rs) {
                r[k] = ls
                i++
            } else {
                r[k] = rs
                j--
            }

            k--
        }
        return r
    }
}